tree/ntuple/v7/src/RPageStorageFile.cxx

```diff
@@ -340,3 +376,130 @@ std::unique_ptr<ROOT::Experimental::Detail::RPageSource> ROOT::Experimental::Det
    clone->fReader = Internal::RMiniFileReader(clone->fFile.get());
    return std::unique_ptr<RPageSourceFile>(clone);
 }
+
+std::unique_ptr<ROOT::Experimental::Detail::RCluster>
+ROOT::Experimental::Detail::RPageSourceFile::LoadCluster(DescriptorId_t clusterId, const ColumnSet_t &columns)
+{
+   fCounters->fNClusterLoaded.Inc();
+
+   const auto &clusterDesc = GetDescriptor().GetClusterDescriptor(clusterId);
+   auto clusterLocator = clusterDesc.GetLocator();
+   auto clusterSize = clusterLocator.fBytesOnStorage;
+   R__ASSERT(clusterSize > 0);
+
+   struct ROnDiskPageLocator {
+      ROnDiskPageLocator() = default;
+      ROnDiskPageLocator(DescriptorId_t c, NTupleSize_t p, std::uint64_t o, std::uint64_t s)
+         : fColumnId(c), fPageNo(p), fOffset(o), fSize(s) {}
+      DescriptorId_t fColumnId = 0;
+      NTupleSize_t fPageNo = 0;
+      std::uint64_t fOffset = 0;
+      std::uint64_t fSize = 0;
+      std::size_t fBufPos = 0;
+   };
+
+   // Collect the page necessary page meta-data and sum up the total size of the compressed and packed pages
+   std::vector<ROnDiskPageLocator> onDiskPages;
+   auto activeSize = 0;
+   for (auto columnId : columns) {
+      const auto &pageRange = clusterDesc.GetPageRange(columnId);
+      NTupleSize_t pageNo = 0;
+      for (const auto &pageInfo : pageRange.fPageInfos) {
+         const auto &pageLocator = pageInfo.fLocator;
+         activeSize += pageLocator.fBytesOnStorage;
+         onDiskPages.emplace_back(ROnDiskPageLocator(
+            columnId, pageNo, pageLocator.fPosition, pageLocator.fBytesOnStorage));
+         ++pageNo;
+      }
+   }
+
+   // Linearize the page requests by file offset
+   std::sort(onDiskPages.begin(), onDiskPages.end(),
+      [](const ROnDiskPageLocator &a, const ROnDiskPageLocator &b) {return a.fOffset < b.fOffset;});
+
+   // In order to coalesce close-by pages, we collect the sizes of the gaps between pages on disk.  We then order
+   // the gaps by size, sum them up and find a cutoff for the largest gap that we tolerate when coalescing pages.
+   // The size of the cutoff is given by the fraction of extra bytes we are willing to read in order to reduce
+   // the number of read requests.  We thus schedule the lowest number of requests given a tolerable fraction
+   // of extra bytes.
+   float maxOverhead = 0.25 * float(activeSize);
```

Comment:
The 0.25 is likely to be a not-necessary-irrelevant tuning parameter (may need to be adjusted base on link/hardware latency and speed and user tolerance for 'extra' cost) ... i.e. may need to be eventually modifiable at run-time.

Reply:
Another related tuning possibility is rounding up or down to fit the device block size.

Reply:
Yep, I added a TODO:

    // TODO(jblomer): Eventually we may want to select the parameter at runtime according to link latency and speed,
    // memory consumption, device block size.